docs/specs/captcha.md

```diff
@@ -1,41 +1,354 @@
+      - [Future use case: Fail-open instead of Fail-close](#future-use-case-fail-open-instead-of-fail-close)
+      - [Future use case: Use risk assessment to determine whether captcha is required](#future-use-case-use-risk-assessment-to-determine-whether-captcha-is-required)
+  * [Audit log](#audit-log)
+  * [Study on Captcha providers](#study-on-captcha-providers)
+    + [Geetest v4](#geetest-v4)
+    + [Geetest v3](#geetest-v3)
+    + [hCaptcha](#hcaptcha)
+    + [reCAPTCHA v2](#recaptcha-v2)
+    + [reCAPTCHA v3](#recaptcha-v3)
+    + [reCAPTCHA Enterprise](#recaptcha-enterprise)
+    + [Cloudflare Turnstile](#cloudflare-turnstile)
+    + [Arkose Labs Bot Manager](#arkose-labs-bot-manager)
+    + [Tencent Captcha](#tencent-captcha)
 
-## Configurations
+## Old configuration
 
-### authgear.yaml
+> The old configuration **IS NOT** used by Authentication Flow.
+> To configure Captcha providers for an Authentication Flow, the new configuration must be used.
+
+See [here](./captcha_legacy.md#configuration) for the documentation of the old configuration.
 
-Before using any workflow with captcha enabled, the captcha provider must be configured in `authgear.yaml`:
+## New Configuration
+
+The section documents the new configuration.
+
+### authgear.yaml
 
 ```yaml
 captcha:
-  provider: cloudflare
+  enabled: true
+  providers:
+  - type: cloudflare
+    alias: cloudflare
+    site_key: "SITE_KEY"
 ```
 
-- `provider`
-  - The captcha provider. Currently, only `cloudflare` is supported.
+- `captcha.enabled`: Boolean. If it is true, the new configuration is used.
+- `captcha.providers`: An array of Captcha provider configuration. The actual shape depends on the `type` property.
+- `captcha.providers.type`: Required. The type of the Captcha provider. Valid values are `cloudflare` and `recaptchav2`.
+- `captcha.providers.alias`: Required. The unique identifier to the Captcha provider. It is used in other parts of the configuration to refer to this particular Captcha provider. For example, the project can configured a number of Captcha providers, but only uses one of them in a particular Authentication Flow.
 
-### authgear.secrets.yaml
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
 
-When the captcha provider is configured, the corresponding secret must exist in `authgear.secrets.yaml`.
+- `site_key`: The site key of Cloudflare Turnstile.
+
+#### `type: recaptchav2`
+
+- `site_key`: The site key of reCAPTCHA v2.
+
+### authgear.secrets.yaml
 
 ```yaml
 - data:
-    secret: YOUR_TURNSTILE_SECRET_KEY
-  key: captcha.cloudflare
+    items:
+    - type: cloudflare
+      alias: cloudflare
+      secret: TURNSTILE_SECRET_KEY
+  key: captcha.providers
+```
+
+- `items.type`: Required. It is the same as `captcha.providers.type`.
+- `items.alias`: Required. It is the same as `captcha.providers.alias`.
+
+Other fields are specific to `type`.
```

Comment:
question: For implementation convenience, should we group these type-specific fields in a field call `configs` (open to suggestion)

Referenced how AuthenticationFlowObjectFlowRoot implemented `GetSteps()` for different Step variation of auth flow

```golang
type AuthenticationFlowObjectFlowRoot interface {
	AuthenticationFlowObject
	GetName() string
	GetSteps() []AuthenticationFlowObject
}
```

Reply:
Not sure what you meant. Can you provide me some examples? (like your own proposal on the shape)

Reply:
How should I implement provider-specific fields?
Please find my attempt below, with some problems I ran into 🙏 

```golang
package config

var _ = Schema.Add("CaptchaConfig", `
{
	"type": "object",
	"additionalProperties": false,
	"properties": {
		"enabled": { "type": "boolean" },
		"providers": { "type": "array", "$ref": "#/$defs/CaptchaProvider" }
	}
}
`)

var _ = Schema.Add("CaptchaProvider", `
{
	"type": "string",
	"enum": ["cloudflare"]
}
`)

type CaptchaConfig struct {
	Enabled   bool                `json:"enabled,omitempty"`
	Providers []CaptchaProviderObject `json:"providers,omitempty"`
}

type CaptchaProviderObject interface {
	IsCaptchaProviderObject()
}

type CaptchaProviderType string

const (
	CaptchaProviderTypeCloudflare CaptchaProviderType = "cloudflare"
	CaptchaProviderTypeRecaptchaV2 CaptchaProviderType = "recaptcha_v2"
)

type CaptchaProviderAlias string

const (
	CaptchaProviderAliasCloudflare CaptchaProviderAlias = "cloudflare"
	CaptchaProviderAliasRecaptchaV2 CaptchaProviderAlias = "recaptcha_v2"
)

type CaptchaProvider struct {
	Type CaptchaProviderType `json:"type,omitempty"`
	Alias CaptchaProviderAlias `json:"alias,omitempty"`
}

func (p *CaptchaProvider) IsCaptchaProviderObject() {}
func (p *CaptchaProvider) GetType() CaptchaProviderType { return p.Type }
func (p *CaptchaProvider) GetAlias() CaptchaProviderAlias { return p.Alias }

var _ CaptchaProviderRoot = &CaptchaProvider{}

// Problem here, how do I define per-provider config?
// Maybe sth like below?

const (
	CaptchaProviderCloudflare = &CaptchaProvider{Type: CaptchaProviderTypeCloudflare, Alias: CaptchaProviderAliasCloudflare} // add cloudflare-specific fields here?
	CaptchaProviderRecaptchaV2 = &CaptchaProvider{Type: CaptchaProviderTypeRecaptchaV2, Alias: CaptchaProviderAliasRecaptchaV2} // add recaptcha-specific fields here?
)

// but above will throw &CaptchaProvider{…} (value of type *CaptchaProvider) is not constant
// what about var?

var (
	CaptchaProviderCloudflare = &CaptchaProvider{Type: CaptchaProviderTypeCloudflare, Alias: CaptchaProviderAliasCloudflare} // add cloudflare-specific fields here?
	CaptchaProviderRecaptchaV2 = &CaptchaProvider{Type: CaptchaProviderTypeRecaptchaV2, Alias: CaptchaProviderAliasRecaptchaV2} // add recaptcha-specific fields here?
)



```

Reply:
Or maybe we can do it like AuthFlow, make these options as a property instead? Expected usage:

```
// authgear.yaml
captcha:
  cloudfare:
     secret: verysecret
  recaptchav2:
     secret: verysecret
```

Reply:
This comment thread is resolved [here](https://github.com/authgear/authgear-server/pull/4337#discussion_r1639283104)